lib/index.js

```diff
@@ -0,0 +1,134 @@
+const isString = require('is-string');
+const assert = require('assert');
+
+class Rule {
+  constructor(opts) {
+    this.opts = opts;
+  }
+}
+
+function isMap(obj) {
+  return obj instanceof Map;
+}
+
+function isFunction(obj) {
+  return typeof obj === 'function';
+}
+
+function mapFromString(s) {
+  const iterable = Array.from(s).map(c => [c, '']);
+  return new Map(iterable);
+}
+
+Rule.from = candidateRule => {
+  if (candidateRule instanceof Rule) return candidateRule;
+  let newRule;
+  if (isString(candidateRule)) {
+    newRule = new MapRule(mapFromString(candidateRule));
+  } else if (isMap(candidateRule)) {
+    newRule = new MapRule(candidateRule);
+  } else if (isFunction(candidateRule)) {
+    newRule = new FunctionRule(candidateRule);
+  } else {
+    throw new Error(`expected a String, Function, or Map: ${candidateRule}`);
+  }
+  return newRule;
+};
+
+class FunctionRule extends Rule {
+  constructor(fn, opts = { modifiesCharacters: true }) {
+    super(opts);
+    this.fn = fn;
+  }
+
+  apply(s1, s2) {
+    return this.fn(s1, s2);
+  }
+}
+
+// TODO: ideally, MapRule would mixin Map.
+class MapRule extends Rule {
+  constructor(map, opts = { modifiesCharacters: false }) {
+    super(opts);
+    assert(map instanceof Map);
+    this.map = map;
+  }
+
+  apply() {
+    throw new Error('MapRules cannot be directly applied.');
+  }
+
+  entries() {
+    return this.map.entries();
+  }
+
+  keys() {
+    return this.map.keys();
+  }
+}
+
+const capitalization = (s1, s2) => {
+  return [s1.toLowerCase(), s2.toLowerCase()];
+};
+const capitalizationRule = new FunctionRule(capitalization);
+
+const unicodeNormalization = (s1, s2) => {
+  return [s1.normalize('NFC'), s2.normalize('NFC')];
```

Comment:
Unfortunately `String#normalize` is not supported in IE and the Android browser. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize#Browser_compatibility 

Currently we have the `normalize.js` util which tries to use `String#normalize` if available and falls back to manually converting a subset of characters we care about.

Reply:
Good note. https://bundlephobia.com/result?p=unorm@1.4.1 is a problem. My thinking for the browser build is:
- initially use normalize.js approach, and can easily add more mappings as needed (and note this approach/limitations in docs)
- longer term, look into downloading unorm in the background, so it is more likely to become available. 
- unorm maintainer is open to improvements: https://github.com/walling/unorm/issues/17 Seems like there could be low-hanging fruit, and I also wonder if it would be possible to do modular builds -- we don't need the compatibility normalization forms, maybe they account for a chunk of the size.